sdk/core/azure-core/src/main/java/com/azure/core/util/Configuration.java

```diff
@@ -323,21 +365,29 @@ private String getOrLoad(String name) {
      * @return If found the loaded configuration, otherwise null.
      */
     private String load(String name) {
-        String value = loadFromProperties(name);
+        if (sectionName == null) {
+            for (ConfigurationSource source : configurationSources) {
+                String value = source.getValue(name);
+                if (value != null) {
+                    return value;
+                }
+            }
 
-        if (value != null) {
-            return value;
+            return null;
         }
 
-        return loadFromEnvironment(name);
-    }
-
-    String loadFromEnvironment(String name) {
-        return System.getenv(name);
-    }
+        // TODO(configration): this is too hard, confusing and relies on property names conventions
+        // need API or other hint for it. making clients do global lookup is even harder
```

Comment:
A question I'll ask here, is it better for the Configuration to have a sectionName or for the caller to pass an input that includes a section name. For example, right now in the PR it is still using `String` as the only getter method property, what if we added a new API that took a more complex type, let's go with `ConfigurationProperty`.

```java
public final class ConfigurationProperty {
  // Name of the property such as "http-log-detail-level"
  private final String propertyName;

  // Whether this property is able to access a non-sectioned/prefixed property
  // Given case, prefix scope of appconfiguration, with the above if "appconfiguration.http-log-detail-level"
  // didn't exist would it be allowed to use "http-log-detail-level".
  private final boolean allowedToUseGlobalProperty;

  private final String[] prefixes;

  public ConfigurationProperty(String propertyName, boolean allowedToUseGlobalProperty) {
  this(propertyName, allowedToUseGlobalProperty, null);
  }

  ConfigurationProperty(String propertyName, boolean allowedToUseGlobalProperty, String[] prefixes) {
    this.propertyName = propertyName;
    this.allowedToUseGlobalProperty = allowedToUseGlobalProperty;
    this.prefixes = prefixes;
  }

  public ConfigurationProperty withAdditionalPrefix(String prefix) {
    if (prefixes == null) {
      return new ConfigurationProperty(propertyName, allowedToUseGlobalPropery, new String[] { prefix });
    } else {
      String[] newPrefixes = new String[prefixes.length + 1);
      newPrefixes[0] = prefix;
      // Copy other prefixes here
     return new ConfigurationProperty(propertyName, allowedToUseGlobalProperty, newPrefixes);
    }
}
```

With this we can offer common properties as static `ConfigurationProperty`s that will know whether they're able to access scopes above their prefix level. And we'll also allow customers to be able to add additional prefix level as well.

Reply:
thanks for fleshing it out!

I've experimented with ConfigurationProperties and I think they are useful. With the current (very early) iteration, I wanted to see the bare minimum API and also got inspired by [.NET IConfiguration](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.configuration.iconfiguration?view=dotnet-plat-ext-6.0).

One scenario `getSection` makes easy, which is hard with `ConfigurationProperties` is multi-layer sections. `getSection`
`ConfigurationProperties` works fine with two-layer config: `<client>.<property>`, but if I want to do <named-client-instance>.<client>.<http-client>.headers, we'll have to support path in `ConfigurationProperties` and it might be harder.
Sections form this path in a bit more efficient way (per client instead of per property).
 
I don't dismiss `ConfigurationProperties`, just trying to start very small and add things when have to. Agree I need to do more research here.

Reply:
another approach I thought about is to have helpers on top of this API that deal with aliases and also do the local/global stuff, e.g.

```
getLocalOrGlobal(String client, String prop) {
  value = configuration.get(client.prop)
  if (value == null) 
    value = configuration.get(prop)
}
```

Same with aliases. We don't even have to expose them to clients as most complicated logic, all global properties would stay in core. 
Clients should use mostly local properties and no aliases, so they would not even need this complexity or on rare occasions, they can write their own helpers. 

If clients would need this, we can expose helpers or add more structure on top of the string name.

Reply:
chatted with @srnagar about a related thing and he had a brilliant idea of using annotations to describe property names and metadata (e.g. local vs global). It also seems that existing Java configuration solutions do (e.g. [Eclipse Microprofile-config](https://github.com/eclipse/microprofile-config/blob/master/spec/src/main/asciidoc/configexamples.asciidoc) which is what Quarcus configuration is based on). they also support prefixes and aliases - I'll try to reuse some ideas and will also look more into annotations.